ethpm_types/contract_type.py

```diff
@@ -324,9 +324,34 @@ def get_deployment_bytecode(self) -> Optional[HexBytes]:
 
         return None
 
+    @cached_property
+    def identifier_definitions(self) -> Dict[str, str]:
+        """
+        Returns a mapping of the full suite of signatures to selectors/topics/IDs for this
+        contract
+        """
+
+        def get_id(aitem: ABI) -> str:
+            if isinstance(aitem, MethodABI) or isinstance(aitem, ErrorABI):
+                return HexBytes(self._selector_hash_fn(aitem.selector)[:4]).hex()
+            else:
+                assert hasattr(aitem, "selector")
+                return HexBytes(self._selector_hash_fn(aitem.selector)).hex()
+
+        abis_with_selector = [x for x in self.abi if hasattr(x, "selector")]
+        return {x.selector: get_id(x) for x in abis_with_selector}
+
+    @cached_property
+    def identifier_lookup(self) -> Dict[str, str]:
+        """
+        Returns a mapping of the full suite of selectors/topics/IDs of this contract to human
+        readable signature
+        """
+        return {v: k for k, v in self.identifier_definitions.items() if v is not None}
+
     @computed_field(alias="methodIdentifiers")  # type: ignore
     @cached_property
```

Comment:
thought: I wonder if this should no longer be individually cached and instead utilize the definitions from the new properties which _are_ cached.

Reply:
thought: as a further extensions of this, we can have `@property`s for all sub-types (just leave off the `@computed_field` part), e.g. `event_identifiers`, `error_identifiers` etc.

perhaps is a bit extra though so feel free to ignore this one, up to you

Reply:
I thought the `computed_field` decorator was for API compatibility with manifest specs or whatever that's from?

I'm fine with adding other methods that filter down to method/error/event.  Or a filter arg?  Though the pre-existing method kind of forces our hand.

You tell me what you'd like to see and I'll put it together.

Reply:
no, seems to be misunderstanding. i don't mean doing anything with `@computed_field`.

Reply:
I guess my idea would involve using `@cached_property` elsewhere to store the ABIs so nevermind.

Reply:
My concern here is the method signatures are cached twice unnecessarily and I was trying to prevent that... if you can think of any way to do this.

Reply:
To go high up enough in the chain to affect everything, we'd basically wrap `self.abi` which is just a list stored in memory anyway so no net-benefit.  Thought maybe cache `_get_abis()` but that groups by types and isn't usable for what I just added.

Could change `method_identifiers()` to pull from `selector_identifiers()` but there'll still be some data transform going on kind of wanting caching going on.

Maybe we're over-optimizing?

Reply:
Let's maybe revisit, it is not a big concern, just something I noticed.

Reply:
I think I addressed this concern mostly in my latest change.  However, removing the `@cached_property` decorator from `method_identifiers()` broke a bunch of other tests so I just added it back.  :sweat_smile: